spec/rspec/matchers/built_in/yield_spec.rb

```diff
@@ -321,6 +329,16 @@ def each_arg(*args, &block)
       }.to fail_with(/expected given block to yield with arguments, but yielded with no arguments/)
     end
 
+    it 'fails if the matchers match at return time only' do
+      expect {
+        expect { |b|
+          val = [1]
+          _yield_with_args(val, &b)
+          val.clear
+        }.to yield_with_args(be_empty)
+      }.to fail_with(/but yielded with unexpected arguments/)
```

Comment:
As I mentioned above, I think we need spec coverage of the fact that the failure messages shows the args as they were at yield time.  I think your `fail_with` expectation here could be expanded to cover that for the positive expectation case.